python/BioSimSpace/Sandpit/Exscientia/FreeEnergy/_restraint.py

```diff
@@ -99,132 +84,134 @@ def __init__(self, system, restraint_dict, type='Boresch'):
                     raise ValueError(
                         f"restraint_dict['force_constants']['{key}'] must be of type "
                         f"'BioSimSpace.Types.Energy'/'BioSimSpace.Types.Angle^2'")
-
-            self._content = restraint_dict
-
         else:
             raise NotImplementedError(f'Restraint type {type} not implemented '
                                       f'yet. Only boresch restraint is supported.')
 
+        self._content = restraint_dict
+        self.update_system(system)
+
+    def update_system(self, system):
+        '''Update the system object.'''
+        if not isinstance(system, _System):
+            raise TypeError("'system' must be of type 'BioSimSpace._SireWrappers.System'")
+        else:
+            if self.rest_type == 'boresch':
+                # Check if the ligand atoms are decoupled.
+                # Find the decoupled molecule, assume that only one can be
+                # decoupled.
+                (decoupled_mol,) = system.getDecoupledMolecules()
+                for key in ['l1', 'l2', 'l3']:
+                    atom = self._content['anchor_points'][key]
+                    if not atom._sire_object.molecule().number() == decoupled_mol._sire_object.number():
```

Comment:
Why do you need to loop over l1, 2, and 3, if we expect them all to have the same molecule number, and they are all compared to the same number (decoupled_mol._sire_object.number())?

Would it be more robust and specific to do something like:
```
for key in ['l1', 'l2', 'l3']:
   atom = self._content['anchor_points'][key]
   if not atom in decoupled_mol:
...
```
Could also check the protein anchor atoms generally with `if atom in system`, if I am understanding this correctly: https://github.com/michellab/BioSimSpace/pull/331.